src/frontend/SignatureMismatch.ml

```diff
@@ -170,34 +179,56 @@ let promote es promotions =
       match prom with
-      | RealPromotion when is_int_type emeta.type_ ->
+      | IntToRealPromotion when is_int_type emeta.type_ ->
           Ast.
             { expr= Ast.Promotion (exp, UReal, emeta.ad_level)
             ; emeta= {emeta with type_= promote_array emeta.type_ UReal} }
-      | ComplexPromotion when not (is_complex_type emeta.type_) ->
+      | (IntToComplexPromotion | RealToComplexPromotion)
+        when not (is_complex_type emeta.type_) ->
           { expr= Promotion (exp, UComplex, emeta.ad_level)
           ; emeta= {emeta with type_= promote_array emeta.type_ UComplex} }
       | _ -> exp )
 
-let returntype env name args =
+let promotion_cost p =
+  match p with
+  | None -> 0
+  | RealToComplexPromotion | IntToRealPromotion -> 1
+  | IntToComplexPromotion -> 2
+
+let matching_function env name args =
   (* NB: Variadic arguments are special-cased in the typechecker and not handled here *)
   let name = Utils.stdlib_distribution_name name in
-  Environment.find env name
-  |> List.filter_map ~f:extract_function_types
-  |> List.sort ~compare:(fun (x, _, _, _) (y, _, _, _) ->
-         UnsizedType.compare_returntype x y )
-  (* Check the least return type first in case there are multiple options (due to implicit UInt-UReal conversion), where UInt<UReal *)
-  |> List.fold_until ~init:[]
-       ~f:(fun errors (rt, tys, funkind_constructor, _) ->
-         match check_compatible_arguments 0 tys args with
-         | Ok p -> Stop (Ok (rt, funkind_constructor, p))
-         | Error e -> Continue (((rt, tys), e) :: errors) )
-       ~finish:(fun errors ->
-         let errors =
-           List.sort errors ~compare:(fun (_, e1) (_, e2) ->
-               compare_errors e1 e2 ) in
-         let errors, omitted = List.split_n errors max_n_errors in
-         Error (errors, not (List.is_empty omitted)) )
+  let unique_minimum_promotion ps =
+    let size (_, _, p) =
+      List.fold ~init:0 ~f:(fun acc p -> acc + promotion_cost p) p in
+    let ps =
+      List.sort ~compare:(fun p1 p2 -> Int.compare (size p1) (size p2)) ps in
+    match ps with
+    | (((rt, tys), _, p1) as ans1) :: ((_, _, p2) as ans2) :: _
+      when size ans1 = size ans2 ->
+        Error (Some ((rt, tys), PromotionConflict (p1, p2)))
+    | ans :: _ -> Ok ans
+    | [] -> Error None in
+  let function_types =
+    Environment.find env name
+    |> List.filter_map ~f:extract_function_types
+    |> List.sort ~compare:(fun (x, _, _, _) (y, _, _, _) ->
+           UnsizedType.compare_returntype x y ) in
+  let matches, errors =
+    List.partition_map function_types
+      ~f:(fun (rt, tys, funkind_constructor, _) ->
+        match check_compatible_arguments 0 tys args with
+        | Ok p -> Either.First ((rt, tys), funkind_constructor, p)
+        | Error e -> Second ((rt, tys), e) ) in
+  match unique_minimum_promotion matches with
+  | Ok ((rt, _), funkind_constructor, p) -> Ok (rt, funkind_constructor, p)
+  | Error (Some e) -> Error ([e], true)
```

Comment:
Returning a single signature here is kind of wrong; this particular signature is acceptable but the problem is that there are many such signatures available.
Let's change the return type of `matching_function` so that it can return multiple _matching_ signatures and then handle that as a new error in the typechecker.
```ocaml
type match_result =
  | UniqueMatch of
      UnsizedType.returntype
      * (bool Middle.Fun_kind.suffix -> Ast.fun_kind)
      * promotions list
  | AmbiguousMatch of (UnsizedType.t list * UnsizedType.returntype) list
  | Errors of signature_error list * bool
```